# Activity: Discover what is in your dataset

## Introduction

In this activity, you will discover characteristics of a dataset and use visualizations to analyze the data. This will develop and strengthen your skills in **exploratory data analysis (EDA)** and your knowledge of functions that allow you to explore and visualize data. 

EDA is an essential process in a data science workflow. As a data professional, you will need to conduct this process to better understand the data at hand and determine how it can be used to solve the problem you want to address. This activity will give you an opportunity to practice that process and prepare you for EDA in future projects.

In this activity, you are a member of an analytics team that provides insights to an investing firm. To help them decide which companies to invest in next, the firm wants insights into **unicorn companies**–companies that are valued at over one billion dollars. The data you will use for this task provides information on over 1,000 unicorn companies, including their industry, country, year founded, and select investors. You will use this information to gain insights into how and when companies reach this prestigious milestone and to make recommendations for next steps to the investing firm.

## Step 1: Imports

### Import libraries and packages 

First, import relevant Python libraries and modules. Use the `pandas` library and the `matplotlib.pyplot` module.

In [2]:
# Import libraries and packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Load the dataset into a DataFrame

The dataset provided is in the form of a csv file named `Unicorn_Companies.csv` and contains a subset of data on unicorn companies. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [3]:
# RUN THIS CELL TO IMPORT YOUR DATA.

companies = pd.read_csv("Unicorn_Companies.csv.csv")

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to what you learned about [loading data](https://www.coursera.org/learn/go-beyond-the-numbers-translate-data-into-insight/supplement/MdTG2/reference-guide-import-datasets-using-python) in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the function in the `pandas` library that allows you to read data from a csv file and load the data into a DataFrame.
 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `read_csv()` function from the `pandas` library. 

</details>

## Step 2: Data exploration

### Display the first 10 rows of the data

Next, explore the dataset and answer questions to guide your exploration and analysis of the data. To begin, display the first 10 rows of the data to get an understanding of how the dataset is structured.

In [4]:
# Display the first 10 rows of the data

companies.head(10)

Company Valuation Date Joined                             Industry  \
0     Bytedance     $180B      4/7/17              Artificial intelligence   
1        SpaceX     $100B     12/1/12                                Other   
2         SHEIN     $100B      7/3/18      E-commerce & direct-to-consumer   
3        Stripe      $95B     1/23/14                              Fintech   
4        Klarna      $46B    12/12/11                              Fintech   
5         Canva      $40B      1/8/18         Internet software & services   
6  Checkout.com      $40B      5/2/19                              Fintech   
7     Instacart      $39B    12/30/14  Supply chain, logistics, & delivery   
8     JUUL Labs      $38B    12/20/17                    Consumer & retail   
9    Databricks      $38B      2/5/19          Data management & analytics   

            City  Country/Region      Continent  Year Founded Funding  \
0        Beijing           China           Asia          2012     $8B   
1      Hawthorne   United States  North America          2002     $7B   
2       Shenzhen           China           Asia          2008     $2B   
3  San Francisco   United States  North America          2010     $2B   
4      Stockholm          Sweden         Europe          2005     $4B   
5    Surry Hills       Australia        Oceania          2012   $572M   
6         London  United Kingdom         Europe          2012     $2B   
7  San Francisco   United States  North America          2012     $3B   
8  San Francisco   United States  North America          2015    $14B   
9  San Francisco   United States  North America          2013     $3B   

                                    Select Investors  
0  Sequoia Capital China, SIG Asia Investments, S...  
1  Founders Fund, Draper Fisher Jurvetson, Rothen...  
2  Tiger Global Management, Sequoia Capital China...  
3        Khosla Ventures, LowercaseCapital, capitalG  
4  Institutional Venture Partners, Sequoia Capita...  
5  Sequoia Capital China, Blackbird Ventures, Mat...  
6  Tiger Global Management, Insight Partners, DST...  
7  Khosla Ventures, Kleiner Perkins Caufield & By...  
8                            Tiger Global Management  
9  Andreessen Horowitz, New Enterprise Associates...

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to [the content about exploratory data analysis in Python](https://www.coursera.org/learn/go-beyond-the-numbers-translate-data-into-insight/lecture/kfl9b/find-stories-using-the-six-exploratory-data-analysis-practices).

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the function in the `pandas` library that allows you to get a specific number of rows from the top of a DataFrame.
 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `head()` function from the `pandas` library. 

</details>

**Question:** What do you think the "Date Joined" column represents?

- the 'Date Joined' column represents the day when the company stars theirs activities or business around the world

**Question:** What do you think the "Select Investors" column represents?

- the 'Select Investors' column represents the mayoritary actionyst into company

### Assess the size of the dataset

Get a sense of how large the dataset is. The `size` property that DataFrames have can help.

In [5]:
# How large the dataset is

companies.size

10740

**Question:** What do you notice about the size of the dataset?

- the dataframe size is represented by 10740 rows it´s means that are 10740 worldwide unicorns companies whose theirs valuation is almost 

### Determine the shape of the dataset

Identify the number of rows and columns in the dataset. The `shape` property that DataFrames have can help.

In [6]:
# Shape of the dataset

companies.shape

(1074, 10)

**Question:** What do you notice about the shape of the dataset?

- the dataframe shape is 1074 rows (unicorn companies) and 10 columns atributes over companies

### Get basic information about the dataset

To further understand what the dataset entails, get basic information about the dataset, including the data type of values in each column. There is more than one way to approach this task. In this instance, use the `info()` function from `pandas`.

In [ ]:
# Get information

companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1074 entries, 0 to 1073
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Company           1074 non-null   object
 1   Valuation         1074 non-null   object
 2   Date Joined       1074 non-null   object
 3   Industry          1074 non-null   object
 4   City              1058 non-null   object
 5   Country/Region    1074 non-null   object
 6   Continent         1074 non-null   object
 7   Year Founded      1074 non-null   int64 
 8   Funding           1074 non-null   object
 9   Select Investors  1073 non-null   object
dtypes: int64(1), object(9)
memory usage: 84.0+ KB


**Question:** What do you notice about the type of data in the `Year Founded` column? Refer to the output from using `info()` above. Knowing the data type of this variable is helpful because it indicates what types of analysis can be done with that variable, how it can be aggregated with other variables, and so on.

- the 'Year Founded' column is talking about the year of fundation by all unicorn companies, which told us since the data earliest year and the last year data

**Question:** What do you notice about the type of data in the `Date Joined` column? Refer to the output from using `info()` above. Knowing the data type of this variable is helpful because it indicates what types of analysis can be done with that variable and how the variable can be transformed to suit specific tasks.

- yes 'Date Joined' column is an object data type, which we could change to data time data type, using datetime function.

## Step 3: Statistical tests

### Find descriptive statistics

Find descriptive statistics and structure your dataset. The `describe()` function from the `pandas` library can help. This function generates statistics for the numeric columns in a dataset. 

In [ ]:
# Get descriptive statist
companies.describe()

Year Founded
count   1074.000000
mean    2012.895717
std        5.698573
min     1919.000000
25%     2011.000000
50%     2014.000000
75%     2016.000000
max     2021.000000

**Question:** Based on the table of descriptive stats generated above, what do you notice about the minimum value in the `Year Founded` column? This is important to know because it helps you understand how early the entries in the data begin.

- yes, it´s important knowing that the data start in the year of 1919

**Question:** What do you notice about the maximum value in the `Year Founded` column? This is important to know because it helps you understand the most recent year captured by the data. 

- yes, it´s important knowing that the most recent year capturated data is 2021

### Convert the `Date Joined` column to datetime

Use the `to_datetime()` function from the `pandas` library  to convert the `Date Joined` column to datetime. This splits each value into year, month, and date components. This is an important step in data cleaning, as it makes the data in this column easier to use in tasks you may encounter. To name a few examples, you may need to compare "date joined" between companies or determine how long it took a company to become a unicorn. Having "date joined" in datetime form would help you complete such tasks.

In [11]:
# Step 1: Use pd.to_datetime() to convert Date Joined column to datetime 
companies['Date Joined']= pd.to_datetime(companies['Date Joined'])
# Step 2: Update the column with the converted values
companies



Company Valuation Date Joined                         Industry  \
0       Bytedance     $180B  2017-04-07          Artificial intelligence   
1          SpaceX     $100B  2012-12-01                            Other   
2           SHEIN     $100B  2018-07-03  E-commerce & direct-to-consumer   
3          Stripe      $95B  2014-01-23                          Fintech   
4          Klarna      $46B  2011-12-12                          Fintech   
...           ...       ...         ...                              ...   
1069     Zhaogang       $1B  2017-06-29  E-commerce & direct-to-consumer   
1070  Zhuan Zhuan       $1B  2017-04-18  E-commerce & direct-to-consumer   
1071     Zihaiguo       $1B  2021-05-06                Consumer & retail   
1072         Zopa       $1B  2021-10-19                          Fintech   
1073        Zwift       $1B  2020-09-16  E-commerce & direct-to-consumer   

               City  Country/Region      Continent  Year Founded Funding  \
0           Beijing           China           Asia          2012     $8B   
1         Hawthorne   United States  North America          2002     $7B   
2          Shenzhen           China           Asia          2008     $2B   
3     San Francisco   United States  North America          2010     $2B   
4         Stockholm          Sweden         Europe          2005     $4B   
...             ...             ...            ...           ...     ...   
1069       Shanghai           China           Asia          2012   $379M   
1070        Beijing           China           Asia          2015   $990M   
1071      Chongqing           China           Asia          2018    $80M   
1072         London  United Kingdom         Europe          2005   $792M   
1073     Long Beach   United States  North America          2014   $620M   

                                       Select Investors  
0     Sequoia Capital China, SIG Asia Investments, S...  
1     Founders Fund, Draper Fisher Jurvetson, Rothen...  
2     Tiger Global Management, Sequoia Capital China...  
3           Khosla Ventures, LowercaseCapital, capitalG  
4     Institutional Venture Partners, Sequoia Capita...  
...                                                 ...  
1069    K2 Ventures, Matrix Partners China, IDG Capital  
1070                           58.com, Tencent Holdings  
1071  Xingwang Investment Management, China Capital ...  
1072  IAG Capital Partners, Augmentum Fintech, North...  
1073    Novator Partners, True, Causeway Media Partners  

[1074 rows x 10 columns]

In [12]:
# Use .info() to confirm that the update actually took place

companies.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1074 entries, 0 to 1073
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Company           1074 non-null   object        
 1   Valuation         1074 non-null   object        
 2   Date Joined       1074 non-null   datetime64[ns]
 3   Industry          1074 non-null   object        
 4   City              1058 non-null   object        
 5   Country/Region    1074 non-null   object        
 6   Continent         1074 non-null   object        
 7   Year Founded      1074 non-null   int64         
 8   Funding           1074 non-null   object        
 9   Select Investors  1073 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(8)
memory usage: 84.0+ KB


### Create a `Year Joined` column

It is common to encounter situations where you will need to compare the year joined with the year founded. The `Date Joined` column does not just have year—it has the year, month, and date. Extract the year component from the `Date Joined` column and add those year components into a new column to keep track of each company's year joined.

In [ ]:
# Step 1: Use .dt.year to extract year component from Date Joined column
# Step 2: Add the result as a new column named Year Joined to the DataFrame

### YOUR CODE HERE ###


In [ ]:
# Use .head() to confirm that the new column did get added

### YOUR CODE HERE ###


## Step 4: Results and evaluation


### Take a sample of the data

It is not necessary to take a sample of the data in order to conduct the visualizations and EDA that follow. But you may encounter scenarios in the future where you will need to take a sample of the data due to time and resource limitations. For the purpose of developing your skills around sampling, take a sample of the data and work with that sample for the next steps of analysis you want to conduct. Use the `sample()` function for this task.

- Use `sample()` with the `n` parameter set to `50` to randomly sample 50 unicorn companies from the data. Be sure to specify the `random_state` parameter to ensure reproducibility of your work. Save the result to a variable called `companies_sampled`.

In [ ]:
# Sample the data

### YOUR CODE HERE ###


### Visualize the time it took companies to reach unicorn status

Visualize the longest time it took companies to reach unicorn status for each industry represented in the sample. To create a bar plot to visualize this, use the `bar()` function from the `matplotlib.pyplot` module. You'll first need to prepare the data.

In [ ]:
# Prepare data for plotting

### YOUR CODE HERE ###

In [ ]:
# Create bar plot
# with the various industries as the categories of the bars
# and the time it took to reach unicorn status as the height of the bars

### YOUR CODE HERE ###


# Set title

### YOUR CODE HERE ###


# Set x-axis label

### YOUR CODE HERE ###


# Set y-axis label

### YOUR CODE HERE ###


# Rotate labels on the x-axis as a way to avoid overlap in the positions of the text

### YOUR CODE HERE ###


# Display the plot

### YOUR CODE HERE ###


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

To prepare the data for modeling, begin by creating a column that represents the number of years it took each company to reach unicorn status. 

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>
    
To prepare the data for modeling, group the dataframe by industry and get the maximum value in the newly created column for each industry.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

You can use the `plt.bar()` to create the bar plot, passing in the categories and heights of the bars.

You can use `plt.title()`, `plt.xlabel()`, and `plt.ylabel()` to set the title, x-axis label, and y-axis label, respectively. 

You can use `plt.xticks()` to rotate labels on the x-axis of a plot. The parameters `rotation=45, horizontalalignment='right'` can be passed in to rotate the labels by 45 degrees and align the labels to the right. 

You can use `plt.show()` to display a plot.

</details>

**Question:** What do you observe from this bar plot?

[Write your response here. Double-click (or enter) to edit.]

### Visualize the maximum unicorn company valuation per industry

Visualize unicorn companies' maximum valuation for each industry represented in the sample. To create a bar plot to visualize this, use the `bar()` function from the `matplotlib.pyplot` module. Before plotting, create a new column that represents the companies' valuations as numbers (instead of strings, as they're currently represented). Then, use this new column to plot your data.

In [ ]:
# Create a column representing company valuation as numeric data

In [ ]:
# Create bar plot
# with the various industries as the categories of the bars
# and the maximum valuation for each industry as the height of the bars

### YOUR CODE HERE ###


# Set title

### YOUR CODE HERE ###


# Set x-axis label

### YOUR CODE HERE ###


# Set y-axis label

### YOUR CODE HERE ###


# Rotate labels on the x-axis as a way to avoid overlap in the positions of the text  

### YOUR CODE HERE ###


# Display the plot

### YOUR CODE HERE ###


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Repeat the process from the last task, only this time with different variables.
</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the function in the `matplotlib.pyplot` module that allows you to create a bar plot, specifying the category and height for each bar. 

Use the functions in the `matplotlib.pyplot` module that allow you to set the title, x-axis label, and y-axis label of plots. In that module, there are also functions for rotating the labels on the x-axis and displaying the plot. 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `plt.bar()` to create the bar plot, passing in the categories and heights of the bars.

Use `plt.title()`, `plt.xlabel()`, and `plt.ylabel()` to set the title, x-axis label, and y-axis label, respectively. 

Use `plt.xticks()` to rotate labels on the x-axis of a plot. The parameters `rotation=45, horizontalalignment='right'` can be passed in to rotate the labels by 45 degrees and align the labels to the right. 

Use `plt.show()` to display a plot.

</details>

**Question:** What do you observe from this bar plot? 

[Write your response here. Double-click (or enter) to edit.]

## Considerations

**What are some key takeaways that you learned from this lab?**

[Write your response here. Double-click (or enter) to edit.]

**What findings would you share with others?**

[Write your response here. Double-click (or enter) to edit.]

**What recommendations would you share with stakeholders based on these findings?**

[Write your response here. Double-click (or enter) to edit.]

**References**

Bhat, M.A. (2022, March). [*Unicorn Companies*](https://www.kaggle.com/datasets/mysarahmadbhat/unicorn-companies). 



**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.